### ISSUE TITLE
Scrolling on titlebars using touchpad doesn't change windows

### ISSUE THREAD

**Issue by David96** [swaywm/sway#6044]
In sway, using a two finger scroll on the touchpad over the titlebar of a stacked/tabbed layout doesn't change the window. The reason for that lies here:
https://github.com/swaywm/sway/blob/4e02a598a33fb9b306a1ab220438adf82cac07a6/sway/input/seatop_default.c#L713

At least for my touchpad, delta_discrete is always 0. One had to use delta to figure out to which window to change. Now that has some problems.

Simply checking whether delta is >/<0 results in extremely fast scrolling as the resolution is pretty high. Checking whether it's over some threshold isn't optimal either as the delta is a delta to the last scroll event so adding a threshold just waits until you reach a certain speed but if you do you likely scroll more windows than intended.

I'm not sure how to tackle this, to get a good user experience I think we had to keep track of the scroll sequence and sum the deltas to make the window change depend on distance scrolled. I also understand if you don't think it's worth the additional trouble. Or maybe there's an easy solution I just didn't find yet.

I just checked i3 and it did work there, though quite fast. I think this is more a sideeffect of the X input system than an intended behavior but anyway, since in i3 it works I filed it as i3-compat (relevant code seems to be this: https://github.com/i3/i3/blob/b95bab0e34365d7637cc87f11e0b3e5fa0956dfb/src/click.c#L213 so no special handling for touchpads).

For myself I'm currently using this patch:
```diff
diff --git a/sway/input/seatop_default.c b/sway/input/seatop_default.c
index f9eb8c8a..2c7e42bd 100644
--- a/sway/input/seatop_default.c
+++ b/sway/input/seatop_default.c
@@ -709,8 +709,9 @@ static void handle_pointer_axis(struct sway_seat *seat,
 			struct sway_node *active =
 				seat_get_active_tiling_child(seat, tabcontainer);
 			list_t *siblings = container_get_siblings(cont);
+			float delta = event->delta_discrete == 0 ? event->delta : event->delta_discrete;
 			int desired = list_find(siblings, active->sway_container) +
-				round(scroll_factor * event->delta_discrete);
+				round(scroll_factor * delta);
 			if (desired < 0) {
 				desired = 0;
 			} else if (desired >= siblings->length) {
```
Which works okish but for the reasons explained above isn't optimal at all.

**Comment by wmww** [comment]
I personally hope nothing along these lines gets landed without being configurable. I would probably only use it by accident, and it would be annoying. Since you have to move your mouse to the title bar area anyway, seems like scrolling would be more trouble than just clicking on the one you want.

**Comment by nmschulte** [comment]
> Since you have to move your mouse to the title bar area anyway, seems like scrolling would be more trouble than just clicking on the one you want.

I use the title-scroll feature (with my mouse) frequently.  I'm surprised to find that my touchpad didn't work, though this is indicative of differences in how I use Sway on mobile/laptop vs desktop.

**Comment by David96** [comment]
> ... Since you have to move your mouse to the title bar area anyway, seems like scrolling would be more trouble than just clicking on the one you want.

For me personally it's sometimes more easy for my finger to only do vertical movements ie moving up and then adding a second finger to scroll instead of up and then left/right - but that's of course all very much personal preference.

Anyway, I'm fine with either decision of adding that or not, just wanted to bring it up since I personally would prefer to have it.

**Comment by Xyene** [comment]
> At least for my touchpad, delta_discrete is always 0.

It will always be 0 for a touchpad. `delta_discrete` really means "number of ticks on the scroll wheel" ([ref](https://wayland.freedesktop.org/libinput/doc/latest/api/group__event__pointer.html#ga38d12fc6884c9943e261febdb2384b98)), and touchpads don't have those.

> I just checked i3 and it did work there, though quite fast. I think this is more a sideeffect of the X input system than an intended behavior

Right. Tablets automagically work in i3 the same way. I think you'd get ~equivalent behavior by doing `max(1.0, min(-1.0, event->delta))`?

> I'm not sure how to tackle this, to get a good user experience I think we had to keep track of the scroll sequence and sum the deltas to make the window change depend on distance scrolled.

I don't think the distance scrolled is something that's in any way standard across different touchpad models. I don't have a reference for this and might just be wrong, but I recall reading touchpad handling in libinput is very voodoo because there's no standardization on these things.

> since in i3 it works I filed it as i3-compat

It's technically an Xorg-compat ;)

Since it seems a bit controversial, I'll mark it as a proposal. I don't have a horse in this race since I don't have a touchpad device :)

**Comment by David96** [comment]
> I don't think the distance scrolled is something that's in any way standard across different touchpad models. I don't have a reference for this and might just be wrong, but I recall reading touchpad handling in libinput is very voodoo because there's no standardization on these things.

I don't think it has to be perfectly the same on all touchpads, simply should be usable, maybe with a configurable threshold to make up for those differences.

To lay some grounds for the discussion, I opened this draft: #6056

**Comment by GermainZ** [comment]
Alternative, very hacky and CPU-inefficient solution:
```
bindsym --border button4 exec sway-exec-limit focus left
bindsym --border button5 exec sway-exec-limit focus right
```

sway-exec-limit:
```sh
#!/bin/sh

# Only run one instance at a time
pidof -o %PPID -x "$0" && exit 0

swaymsg "$@"
# Modify sleep delay to change sensitivity
sleep 0.2
```
